Add table shadow element to tables

joshuatf opened the pull request:
Fixes #450 

This adds a shadow to all tables with a horizontal overflow until the table has been scrolled to the end.

### Screenshots

<img width="300" alt="screen shot 2018-10-16 at 4 02 22 pm" src="https://user-images.githubusercontent.com/10561050/47043952-f7d61900-d15c-11e8-93c8-e52036aeb21c.png">

### Detailed test instructions:

1.  Open `/wp-admin/admin.php?page=wc-admin#/analytics/revenue`.
2.  Narrow viewport width to make table scrollable.
3.  Scroll to end to make sure shadow is not visible when fully scrolled to the right.
4.  Resize screen to desktop view to confirm that shadow styles are recalculated (hidden) on resize.

joshuatf committed:
Add table shadow element to tables

LevinMedia commented:
Looking good! @joshuatf would you mind updating the background of 

`.woocommerce-table__shadow` 

to

`background: linear-gradient(90deg, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.1), rgba(0,0,0,0.2));`

Thanks!

joshuatf committed:
Update table shadow bg color

joshuatf commented:
@LevinMedia Updated!

Aljullu reviewed (approved):
That makes tables look much better! Some small comments regarding the code and LGTM.

I also wrote a (longer) comment regarding scroll events. I think it doesn't need to be fixed now but it's something we might want to take a look in the future.

Aljullu commented on client/components/table/style.scss:
```diff
@@ -50,6 +51,16 @@
 }
 
+.woocommerce-table__shadow {
+	content: '';
```
I think this can be removed.
```suggestion
```

joshuatf replied:
You're right; this was left from a previous implementation and should not have made it into the PR 😓 - Removed!

Aljullu commented on client/components/table/style.scss:
```diff
@@ -50,6 +51,16 @@
+	top: 0;
+	width: 41px;
+	height: 100%;
+	background: linear-gradient(90deg, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.1), rgba(0, 0, 0, 0.2));
```
The color in the middle shouldn't be necessary:
```suggestion
	background: linear-gradient(90deg, rgba(0, 0, 0, 0), rgba(0, 0, 0, 0.2));
```

joshuatf replied:
Also correct- removed!

Aljullu commented on client/components/table/table.js:
```diff
@@ -191,6 +214,11 @@ class Table extends Component {
 					</tbody>
 				</table>
+				<div
+					ref={ this.tableShadow }
```
Is this ref used somewhere else?

joshuatf replied:
No longer needed- removed!

Aljullu commented on client/components/table/table.js:
```diff
@@ -84,6 +95,17 @@ class Table extends Component {
+		const scrolledToEnd = table.scrollWidth - table.scrollLeft <= table.offsetWidth;
+		this.setState( {
+			shadowStyles: {
+				transform: 'translateX(' + this.container.current.scrollLeft + 'px)',
```
Updating CSS positions with JavaScript is a bit problematic. I recorded this in my PC:
![peek 2018-10-17 16-16](https://user-images.githubusercontent.com/3616980/47094736-0e977100-d22c-11e8-8447-590d663b8d17.gif)

It's hard to see but there is one frame in which the scroll has already happened but the shadow position hasn't been updated yet:
![peek 2018-10-17 16-16](https://user-images.githubusercontent.com/3616980/47094771-1fe07d80-d22c-11e8-8f04-05d08240f870.png)

I imagine that will happen more often in low-end phones.

I have been doing tests with this PR and actually that's a tricky CSS issue, the only way I could find to avoid having to update the position with JavaScript was creating a wrapper `div` around the table which would contain the shadow absolutely positioned to the right. Since that `div` would not scroll, the shadow would always be aligned to the right without having to set its position with JavaScript. But that means polluting the markup, so I'm not sure it's the best solution either.

Given that it's a small visual glitch and there isn't any perfect solution, I'm ok with merging this PR as it is now, but wanted to point it out in case somebody has any other idea. :slightly_smiling_face:

joshuatf replied:
Ah, I thought this might happen.  

I updated this in https://github.com/woocommerce/wc-admin/pull/546/commits/6e16f258c2de953dc11ef9b170a1839e36ebdd8e to be relative to the container table body which should let us position it without constantly updating the transform.  This required a small change to how the pagination is positioned on tables to go above the shadow, but I think it works well across all tables.

Let me know if that looks better on your end!

Aljullu replied:
Right, that was a clever way to fix that. The problem is solved now and it works like a charm!

I have one last suggestion, what would you think about getting rid of `<div className="woocommerce-table__shadow">` and implementing it as a pseudo-element that is visible/hidden depending on a class applied to `woocommerce-table__table`?

I think now it's do-able and this way we will keep the markup simple.

joshuatf replied:
That seems like a good idea.  Pushed in https://github.com/woocommerce/wc-admin/pull/546/commits/159fde6428ee2c10fb87b7f5dc2991743a61df2b

Aljullu replied:
Looks really good to me, now. :clap:

joshuatf committed:
Remove redundant styles

joshuatf committed:
Remove unused ref

joshuatf committed:
Update shadow relative to container table body

Use the container body so we don't have to update the transform on scroll for the table shadow.  Prevents jankiness on low-end devices

joshuatf committed:
Move table shadow to pseudo element
